sympy/solvers/solveset.py

```diff
@@ -1329,13 +1329,14 @@ def _invert_modular(modterm, rhs, n, symbol):
             return _invert_modular(Mod(h, m), Mod(x_indep_term, m), n, symbol)
 
     if a.is_Pow:
+        from sympy.utilities.misc import as_int
         # base**expo = a
         base, expo = a.args
         if expo.has(symbol) and not base.has(symbol):
             # remainder -> solution independent of n of equation.
             # m, rhs are made coprime by dividing number_gcd(m, rhs)
             try:
-                remainder = discrete_log(m / number_gcd(m, rhs), rhs, a.base)
+                remainder = discrete_log(m / number_gcd(as_int(m), as_int(rhs)), rhs, a.base)
```

Comment:
The `as_int` function is intended for preparsing inputs provided by the user to a public function where the user might pass any `int`-like object including say a `numpy.int32` or something.

This function `_invert_modular` is an internal function that will only ever be called with symbolic sympified types (i.e. Expr). If we know that `m` and `rhs` are `Expr` then there is only one case in which `as_int` could succeed which is if the argument is `Integer`. In that case the `.p` attribute gets the value as a plain `int`. I would instead do this like:
```python
if not m.is_Integer and rhs.is_Integer and a.base.is_Integer:
    return modterm, rhs

mdiv = m.p // number_gcd(m.p, rhs.p)
try:
    remainder = discrete_log(mdiv, rhs.p, a.base.p)
except ValueError: # log does not exist.
    return modterm, rhs
```
It is better to use explicit `if` checks rather than depending on catching exceptions. This way it is explicit that we only take this branch when all quantities are integers and any `ValueError` can only come from `discrete_log` (which should ideally use a different exception from `ValueError`).

Reply:
Thank you. I have fixed it that way.